src/math/interpolator.hh

```diff
@@ -0,0 +1,1168 @@
+      // to solve the linear system in
+      // http://mathworld.wolfram.com/CubicSpline.html
+      // under different boundary conditions. We describe with s the spline
+      // function and with y the targeted function for interpolation.
+      //
+      // Reference:
+      // https://en.wikipedia.org/wiki/Tridiagonal_matrix_algorithm
+
+      // clamped boundary conditions when first derivative for boundaries is
+      // known s'(x_0) = y'(x_0), s'(x_n) = y'(x_n)
+      inline void compute_second_derivative(const Vector_Ref & yv, double yd1,
+                                            double ydn) {
+        // Bad xa input to routine splint
+        assert(this->h != 0.0);
+        int n{static_cast<int>(yv.size())};
+        Vector_t y2 = Vector_t::Zero(n);
+        Vector_t u = Vector_t::Zero(n);
+        double p;
+
+        // forward sweeping
+        y2(0) = -0.5;
+        u(0) = (3.0 / this->h) * ((yv(1) - yv(0)) / this->h - yd1);
+        for (int i{1}; i < n - 1; i++) {
+          p = 0.5 * y2(i - 1) + 2.0;
+          y2(i) = -0.5 / p;
+          u(i) = (yv(i + 1) - 2 * yv(i) + yv(i - 1)) / this->h;
+          u(i) = (3.0 * u(i) / this->h - 0.5 * u(i - 1)) / p;
+        }
+
+        // back substitution
+        p = 0.5;  // qn=0.5 but we just reuse p because it is not needed anyway
+        u(n - 1) = (3.0 / this->h) * (ydn - (yv(n - 1) - yv(n - 2)) / this->h);
+        y2(n - 1) = (u(n - 1) - p * u(n - 2)) / (p * y2(n - 2) + 1.0);
+        for (int k{n - 2}; k >= 0; k--) {
+          y2(k) = y2(k) * y2(k + 1) + u(k);
+        }
+        this->second_derivative_h_sq_6 = y2 * this->h_sq_6;
+      }
+
+      // natural/simple boundary conditions s''(x_0) = s''(x_n) = 0
+      inline void compute_second_derivative(const Vector_Ref & yv) {
+        // Bad xa input to routine splint
+        assert(this->h != 0.0);
+        int n{static_cast<int>(yv.size())};
+        Vector_t y2 = Vector_t::Zero(n);
+        Vector_t u = Vector_t::Zero(n);
+        double p;
+
+        // forward sweeping
+        y2(0) = 0.0;
+        u(0) = 0.0;
+        for (int i{1}; i < n - 1; i++) {
+          p = 0.5 * y2(i - 1) + 2.0;
+          y2(i) = -0.5 / p;
+          u(i) = (yv(i + 1) - 2 * yv(i) + yv(i - 1)) / this->h;
+          u(i) = (3.0 * u(i) / this->h - 0.5 * u(i - 1)) / p;
+        }
+
+        // back substitution
+        y2(n - 1) = 0.0;  // (u(n-1)-p*u(n-2))/(p*y2(n-2)+1.0);
+        for (int k{n - 2}; k > 0; k--) {
+          y2(k) = y2(k) * y2(k + 1) + u(k);
+        }
+        y2(0) = 0.0;  // y2(0)*y2(1)+u(0);
+        this->second_derivative_h_sq_6 = y2 * this->h_sq_6;
+      }
+
+      inline double raw_interpolate(const Vector_Ref & xx,
+                                    const Vector_Ref & yy, const int & j1,
+                                    const double & x) {
+        assert(this->h != 0.0);
+        const int klo{j1}, khi{j1 + 1};
+        // a+b=1
```

Comment:
?

Reply:
added comment:
        // percentage of x to grid cell end, because a+b = 1 we can simplify
        // the computation
hope it is clear now